src/types/redis_bloom_chain.cc

```diff
@@ -0,0 +1,198 @@
+  if (!found) {
+    *ret = 1;
+    block_split_bloom_filter.InsertHash(h);
+    auto batch = storage_->GetWriteBatchBase();
+    batch->Put(bf_key, block_split_bloom_filter.GetData());
+    return storage_->Write(storage_->DefaultWriteOptions(), batch->GetWriteBatch());
+  }
+
+  return rocksdb::Status::OK();
+}
+
+rocksdb::Status BloomChain::bloomCheck(const Slice &bf_key, const std::string &item, int *ret) {
+  LatestSnapShot ss(storage_);
+  rocksdb::ReadOptions read_options;
+  read_options.snapshot = ss.GetSnapShot();
+  std::string bf_data;
+  rocksdb::Status s = storage_->Get(read_options, bf_key, &bf_data);
+  if (!s.ok()) return s;
+  BlockSplitBloomFilter block_split_bloom_filter;
+  block_split_bloom_filter.Init(std::move(bf_data));
+
+  uint64_t h = BlockSplitBloomFilter::Hash(item.data(), item.size());
+  bool found = block_split_bloom_filter.FindHash(h);
+  *ret = found ? 1 : 0;
+
+  return rocksdb::Status::OK();
+}
+
+rocksdb::Status BloomChain::Reserve(const Slice &user_key, uint32_t capacity, double error_rate, uint16_t expansion) {
+  std::string ns_key = AppendNamespacePrefix(user_key);
+
+  LockGuard guard(storage_->GetLockManager(), ns_key);
+  BloomChainMetadata bloom_chain_metadata;
+  rocksdb::Status s = getBloomChainMetadata(ns_key, &bloom_chain_metadata);
+  if (!s.ok() && !s.IsNotFound()) return s;
+  if (!s.IsNotFound()) {
+    return rocksdb::Status::InvalidArgument("the key already exists");
+  }
+
+  return createBloomChain(ns_key, error_rate, capacity, expansion, &bloom_chain_metadata);
+}
+
+rocksdb::Status BloomChain::Add(const Slice &user_key, const Slice &item, int *ret) {
+  std::string ns_key = AppendNamespacePrefix(user_key);
+  LockGuard guard(storage_->GetLockManager(), ns_key);
+
+  BloomChainMetadata metadata;
+  rocksdb::Status s = getBloomChainMetadata(ns_key, &metadata);
+
+  if (s.IsNotFound()) {
+    s = createBloomChain(ns_key, kBFDefaultErrorRate, kBFDefaultInitCapacity, kBFDefaultExpansion, &metadata);
+  }
+  if (!s.ok()) {
+    return s;
+  }
+
+  std::vector<std::string> bf_key_list;
+  getBFKeyList(ns_key, metadata, &bf_key_list);
+
+  auto batch = storage_->GetWriteBatchBase();
+  WriteBatchLogData log_data(kRedisBloomFilter, {"insert"});
+  batch->PutLogData(log_data.Encode());
+
+  int check_index = metadata.n_filters - 1;
+  std::string item_string = item.ToString();
+
+  // check
+  for (; check_index >= 0; --check_index) {  // TODO: to test which direction for searching is better
+    s = bloomCheck(bf_key_list[check_index], item_string, ret);
+    if (!s.ok()) return s;
+    if (*ret == 1) {
+      break;
+    }
+  }
+
+  // insert
+  if (check_index < 0) {
+    if (metadata.size + 1 > metadata.GetCapacity()) {  // TODO: scaling would be supported later
+      return rocksdb::Status::Aborted("filter is full");
+    }
+    s = bloomAdd(bf_key_list.back(), item_string, ret);
+    if (!s.ok()) return s;
+    metadata.size += *ret;
+  }
+
+  std::string sb_chain_metadata_bytes;
```

Comment:
Only need to `Put` when `check_index < 0`?

Reply:
Sure, only need to `Put` when `metadata` change. But if I reuse this functions in `MADD`, I think it would be better to leave the `Put` outside and only `Put` once when `metadata.size` change.

Reply:
Ok, you can support `INFO` in next patch. It's easy and it can help us debug:
https://github.com/RedisBloom/RedisBloom/blob/21a2620e75873353fead8c5d70950d3791e36b18/docs/commands/bf.info.md#L4